pallets/elections/src/impls.rs

```diff
@@ -293,6 +299,45 @@ where
 
         T::ValidatorRewardsHandler::add_rewards(rewards);
     }
+
+    fn calculate_underperforming_validators() {
+        let thresholds = CurrentEraCommitteeKickOutThresholds::<T>::get();
+        let current_committee = T::SessionInfoProvider::current_committee();
+        for validator in current_committee {
+            let underperformance = match SessionValidatorBlockCount::<T>::try_get(&validator) {
+                Ok(block_count) => block_count <= thresholds.block_count_threshold,
+                Err(_) => true,
+            };
+            if underperformance {
+                Self::mark_validator_underperformance(&thresholds, &validator)
+            }
+        }
+    }
+
+    fn mark_validator_underperformance(
+        thresholds: &CommitteeKickOutThresholds,
+        validator: &T::AccountId,
+    ) {
+        UnderperformedValidatorSessionCount::<T>::mutate(&validator, |count| {
+            *count += 1;
+        });
+        let counter = UnderperformedValidatorSessionCount::<T>::get(&validator);
+        if counter >= thresholds.underperformed_session_count_threshold {
```

Comment:
Is `UnderperfromedValidatorSessionCount` resetting its values after automatically? I.e. do we initialize a new instance of it every few sessions? I'm worried that if we don't then it will be too strict long-term. We should probably "reset" this counter if validator performed correctly in `n out of M` recent sessions. I.e. I don't want to keep being worried whether I get kicked out for missing a block b/c months ago I had a bad streak.

Reply:
See my proposal below to @DamianStraszak 's post.

Reply:
Actually, now done